internal/api/token_refresh.go

```diff
@@ -60,39 +60,51 @@ func (a *API) RefreshTokenGrant(ctx context.Context, w http.ResponseWriter, r *h
 		}
 	}
 
-	if token.Revoked {
-		a.clearCookieTokens(config, w)
-		// For a revoked refresh token to be reused, it has to fall within the reuse interval.
+	// basic checks above passed, now we need to serialize access to the
+	// session in a transaction so that there's no concurrent modification
 
-		reuseUntil := token.UpdatedAt.Add(
-			time.Second * time.Duration(config.Security.RefreshTokenReuseInterval))
+	var tokenString string
+	var newTokenResponse *AccessTokenResponse
 
-		if time.Now().After(reuseUntil) {
-			// not OK to reuse this token
+	err = db.Transaction(func(tx *storage.Connection) error {
+		user, token, _, terr := models.FindUserWithRefreshToken(db, params.RefreshToken, true /* forUpdate */)
```

Comment:
Would this be performed outside the transaction?

Reply:
It is below once the session has been locked from concurrent access.

Reply:
I'm not entirely sure what you mean, but I was referring to the fact that `models.FindUserWithRefreshToken` is called with `db` here instead of `tx` from the newly created transaction above, which makes me think that the user could be locked outside of the current transaction. Also, is there a reason to create a nested transaction below?